packages/relay/src/services/ReportService.ts

```diff
@@ -280,6 +295,389 @@ export class ReportService {
+            repUserType === undefined
+        ) {
+            throw InvalidParametersError
+        }
+        if (!Object.values(RepUserType).includes(repUserType)) {
+            throw InvalidParametersError
+        }
+    }
+
+    private async getReport(
+        db: DB,
+        reportId: string,
+        repUserType: RepUserType,
+        nullifier?: string
+    ) {
+        if (repUserType === RepUserType.VOTER) {
+            if (!nullifier) throw InvalidParametersError
+            return this.findReportWithNullifier(
+                db,
+                reportId,
+                nullifier,
+                ReportStatus.WAITING_FOR_TRANSACTION
+            )
+        } else {
+            const whereClause = this.getReportWhereClause(reportId)
+            const report = await db.findOne('ReportHistory', {
+                where: whereClause,
+            })
+            if (!report) throw ReportNotExistError
+            return report
+        }
+    }
+
+    private getReportWhereClause(reportId: string) {
+        return {
+            reportId,
+            status: ReportStatus.WAITING_FOR_TRANSACTION,
+        }
+    }
+
+    private checkReputationClaim(
+        report: ReportHistory,
+        repUserType: RepUserType,
+        direction: ReputationDirection,
+        nullifier?: string
+    ) {
+        if (direction === ReputationDirection.NEGATIVE) {
+            if (
+                repUserType === RepUserType.REPORTER &&
+                report.reportorClaimedRep
+            )
+                throw UserAlreadyClaimedError
+            if (
+                repUserType === RepUserType.POSTER &&
+                report.respondentClaimedRep
+            )
+                throw UserAlreadyClaimedError
+        } else {
+            if (
+                repUserType === RepUserType.REPORTER &&
+                report.reportorClaimedRep
+            )
+                throw UserAlreadyClaimedError
+            if (repUserType === RepUserType.VOTER) {
+                if (!nullifier) throw InvalidParametersError
+                if (!report) throw InvalidReportNullifierError
+                if (
+                    report.adjudicatorsNullifier?.some(
+                        (adj: any) => adj.nullifier === nullifier && adj.claimed
+                    )
+                ) {
+                    throw UserAlreadyClaimedError
+                }
+            }
+        }
+    }
+
+    private getReputationChange(
+        direction: ReputationDirection,
+        repUserType: RepUserType
+    ): RepChangeType {
+        if (direction === ReputationDirection.NEGATIVE) {
+            return repUserType === RepUserType.REPORTER
+                ? RepChangeType.FAILED_REPORTER_REP
+                : RepChangeType.POSTER_REP
+        } else {
+            if (repUserType === RepUserType.VOTER) {
+                return RepChangeType.VOTER_REP
+            }
+            return RepChangeType.REPORTER_REP
+        }
+    }
```

Comment:
This implementation will return wrong value below these situations:
- `direction` is `ReputationDirection.NEGATIVE` and `repUserType` is `RepUserType.Voter`
- `direction` is `ReputationDirection.POSITIVE` and `repUserType` is `RepUserType.POSTER`

Reply:
fixed